### PR TITLE
FlutterEngineGroup memory test: added precache before executing gradle.

### PR DESCRIPTION
**Pull request by gaaclarke**
issue: https://github.com/flutter/flutter/issues/72026

This test is still failing for a new reason: https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket.appspot.com/8855695600526402512/+/steps/run_flutter_engine_group_performance/0/stdout

```
[flutter_engine_group_performance] [STDOUT] stdout: Welcome to Gradle 6.5!
[flutter_engine_group_performance] [STDOUT] stdout: 
[flutter_engine_group_performance] [STDOUT] stdout: Here are the highlights of this release:
[flutter_engine_group_performance] [STDOUT] stdout:  - Experimental file-system watching
[flutter_engine_group_performance] [STDOUT] stdout:  - Improved version ordering
[flutter_engine_group_performance] [STDOUT] stdout:  - New samples
[flutter_engine_group_performance] [STDOUT] stdout: 
[flutter_engine_group_performance] [STDOUT] stdout: For more details see https://docs.gradle.org/6.5/release-notes.html
[flutter_engine_group_performance] [STDOUT] stdout: 
[flutter_engine_group_performance] [STDOUT] stdout: To honour the JVM settings for this build a new JVM will be forked. Please consider using the daemon: https://docs.gradle.org/6.5/userguide/gradle_daemon.html.
[flutter_engine_group_performance] [STDOUT] stdout: Daemon will be stopped at the end of the build stopping after processing
[flutter_engine_group_performance] [STDOUT] stdout: > Task :app:preBuild UP-TO-DATE
[flutter_engine_group_performance] [STDOUT] stdout: > Task :app:preReleaseBuild UP-TO-DATE
[flutter_engine_group_performance] [STDOUT] stdout: 
[flutter_engine_group_performance] [STDOUT] stdout: > Task :flutter:compileFlutterBuildRelease
[flutter_engine_group_performance] [STDOUT] stderr: 
[flutter_engine_group_performance] [STDOUT] stderr: Invalid argument(s): Cannot find executable for /opt/s/w/ir/k/flutter sdk/bin/cache/artifacts/engine/android-arm-release/linux-x64/gen_snapshot.
[flutter_engine_group_performance] [STDOUT] stderr: #0      _getExecutable (package:process/src/interface/local_process_manager.dart:124:5)
[flutter_engine_group_performance] [STDOUT] stderr: #1      LocalProcessManager.start (package:process/src/interface/local_process_manager.dart:40:30)
[flutter_engine_group_performance] [STDOUT] stderr: #2      ErrorHandlingProcessManager.start.<anonymous closure> (package:flutter_tools/src/base/error_handling_io.dart:744:24)
[flutter_engine_group_performance] [STDOUT] stderr: #3      _run (package:flutter_tools/src/base/error_handling_io.dart:536:20)
[flutter_engine_group_performance] [STDOUT] stderr: #4      ErrorHandlingProcessManager.start (package:flutter_tools/src/base/error_handling_io.dart:734:12)
[flutter_engine_group_performance] [STDOUT] stderr: #5      _DefaultProcessUtils.start (package:flutter_tools/src/base/process.dart:474:28)
[flutter_engine_group_performance] [STDOUT] stderr: #6      _DefaultProcessUtils.stream (package:flutter_tools/src/base/process.dart:493:35)
[flutter_engine_group_performance] [STDOUT] stderr: #7      GenSnapshot.run (package:flutter_tools/src/base/build.dart:75:26)
[flutter_engine_group_performance] [STDOUT] stderr: #8      AOTSnapshotter.build (package:flutter_tools/src/base/build.dart:209:56)
[flutter_engine_group_performance] [STDOUT] stderr: #9      AndroidAot.build (package:flutter_tools/src/build_system/targets/android.dart:235:52)
```

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by gaaclarke** [comment]
Discussed offline with @jmagman.  While this may fix my issue, there might be another bug in the add-to-app tooling since it would be expected that the artifacts would be downloaded with the steps: `cd $MODULE; flutter pub get; cd $ANDROID; ./gradlew assembleRelease`.

File: dev/devicelab/bin/tasks/flutter_engine_group_performance.dart [hunk 1]
```diff
@@ -59,6 +59,8 @@ Future<TaskResult> _doTest() async {
     final String gradlewExecutable =
         Platform.isWindows ? '.\\$gradlew' : './$gradlew';
     final String flutterPath = path.join(flutterDirectory, 'bin', 'flutter');
+    await utils.eval(flutterPath, <String>['precache', '--linux', '--android'],
```
Comment (jmagman) [inline dev/devicelab/bin/tasks/flutter_engine_group_performance.dart:62]: I don't think you need `--linux`, that is for targeting linux, not running on it.

**Review by jmagman — COMMENTED** [review]
> Discussed offline with @jmagman. While this may fix my issue, there might be another bug in the add-to-app tooling since it would be expected that the artifacts would be downloaded with the steps: `cd $MODULE; flutter pub get; cd $ANDROID; ./gradlew assembleRelease`.

I wouldn't expect that to download artifacts, but a `flutter run` or `flutter build` targeting an Android device would.

This approach looks good to me once you remove `--linux`.

**Comment by gaaclarke** [comment]
> I wouldn't expect that to download artifacts

That is the steps we've advertised for our add-to-app samples: https://github.com/flutter/samples/tree/master/add_to_app/fullscreen#tldr

Also, option B for our documentation doesn't ever ask someone to build the module before you run the host project:
https://flutter.dev/docs/development/add-to-app/android/project-setup#option-b---depend-on-the-modules-source-code

**Review by jmagman — APPROVED** [review]
LGTM

**Comment by jmagman** [comment]
> > I wouldn't expect that to download artifacts
> 
> That is the steps we've advertised for our add-to-app samples: https://github.com/flutter/samples/tree/master/add_to_app/fullscreen#tldr
> 
> Also, option B for our documentation doesn't ever ask someone to build the module before you run the host project:
> https://flutter.dev/docs/development/add-to-app/android/project-setup#option-b---depend-on-the-modules-source-code

You don't need to build the module per se, just run something in any project (build, run, whatever) that targets android to get the android artifacts at least once after a `flutter upgrade`.  On the iOS side at least if you don't have them downloaded the host side `pod install` will fail with a message telling you to run `precache`.  We see this a lot more than most developers who stay on stable, and only upgrade every few months.

https://github.com/flutter/flutter/blob/7711c1ef8f478c03d12d72f790c625a194cc93b9/packages/flutter_tools/templates/module/ios/library/Flutter.tmpl/podhelper.rb.tmpl#L40-L41

Sounds like a good thing to mention in the docs.

**Pull request merged**